### ISSUE TITLE
imshow origin and contour origin behave differently

### ISSUE THREAD

**Issue by anntzer** [matplotlib/matplotlib#17046]
### Bug report

**Bug summary**

`imshow(t, origin="upper")` draws `t[0, 0]` around `x=0, y=0` on the axes but inverts the y axis to go from top to bottom.  On the other hand, `contour(t, origin="upper")` draws the point corresponding to `t[0, 0]` at `x=0, y=len(t)` (modulo off-by-1's that don't matter here), but keeps the axis orientation.  So e.g.
```
t = np.add.outer(np.arange(10)**2, np.arange(10)**2)
origin = "upper"
plt.imshow(t, origin=origin)
plt.contour(t, origin=origin)
```
(which one may think is necessary to write code that works for both origin="upper" and origin="lower") results in the incorrect
![test](https://user-images.githubusercontent.com/1322974/78563148-cce0f880-781a-11ea-93d8-45ce1076fe90.png)

At this point I'm not even sure we can really do anything to change that behavior due to backcompat concerns, but at least this could perhaps 1) be documented better in the docstring of contour and 2) be mentioned in https://matplotlib.org/tutorials/intermediate/imshow_extent.html?  (I haven't even checked how `extent` interacts with contour origin, but it's probably lots of fun too.)

**Matplotlib version**

  * Operating system: any
  * Matplotlib version: head
  * Matplotlib backend (`print(matplotlib.get_backend())`): any
  * Python version: any
  * Jupyter version (if applicable):
  * Other libraries:

**Comment by jklymak** [comment]
The origin stuff should all be deprecated in favour of folks flipping the axis.

**Comment by mcsshep** [comment]
I believe that the cause of this problem is the following excerpt from contour.py, which does not do what its comment claims:

        # Match image behavior:
        if self.extent is None:
            x0, x1, y0, y1 = (0, Nx, 0, Ny)
        else:
            x0, x1, y0, y1 = self.extent

To match the behavior of image.py (ie. imshow), as the comment says it should, I believe that the above code should be modified to do the equivalent of the following excerpt from image.py:

    def get_extent(self):
        """Return the image extent as tuple (left, right, bottom, top)."""
        if self._extent is not None:
            return self._extent
        else:
            sz = self.get_size()
            numrows, numcols = sz
            if self.origin == 'upper':
                return (-0.5, numcols-0.5, numrows-0.5, -0.5)
            else:
                return (-0.5, numcols-0.5, -0.5, numrows-0.5)

Note that when origin is 'lower', the extent that contour uses in the above excerpt is off by half a pixel compared to imshow, and a completely innapropriate value is used for the default extent if origin is 'upper'.

For now, a way to work around this issue is to always remember to set the extent argument of the contour function. In my case, I encountered this issue while using origin='lower' in both imshow and contour, which resulted in an obvious offset of half a pixel between the contours and the imshow pixels.

**Comment by tacaswell** [comment]
Still a problem.

I agree with the analysis in https://github.com/matplotlib/matplotlib/issues/17046#issuecomment-876174514

I disagree we should remove the origin behavior as it really specifies if your data is top-down or bottom-up in its bounding box.

**Comment by jklymak** [comment]
> I disagree we should remove the origin behavior as it really specifies if your data is top-down or bottom-up in its bounding box.

I think its unnecessary compared to `np.flipud(Z)`

**Comment by anntzer** [comment]
> I think its unnecessary compared to np.flipud(Z)

This would make the interactive cursor y values incorrect.  (Concretely, the y values can e.g. correspond to physical pixel rows on a camera so you do want to have the correct y value; and you also want to display the image right side up, of course.)

**Comment by jklymak** [comment]
Yes - but we only need two of "origin", "extent" and "set_ylim"

**Comment by timhoffm** [comment]
There is multiple things here:

### TL;DR 

- The default behavior between `imshow()` and `contour()` is different. But it's correctly documented and justifiable -> no action needed.
- The explicit "upper"/"lower" behavior places the values at unreasonable coordinates. We should come up with a migration strategy for this - I assume this only affects the cases without `X, Y` and without `extent` (t.b.c.), i.e. when the user did not provide any explicit coordinate context for the data. Hopefully, the breakage is limited for these cases. I can hardly imagine a case where the user relies on our awkward positioning. Likely the users either do not care for coordinates or they didn't realize they were off.


![image](https://github.com/matplotlib/matplotlib/assets/2836374/cda0da98-295a-4a82-b22d-1c769e6ba4ca)

<details>
  <summary>Code for reference</summary>

  ```
  t = np.add.outer(np.arange(10)**2, np.arange(10)**2)
  f, axs = plt.subplots(3, 2,  figsize=(6, 10))
  axs[0, 0].set_ylabel('origin="default"')
  axs[0, 0].imshow(t)
  axs[0, 0].set_title(f"ylim={axs[0, 0].get_ylim()}")
  axs[0, 1].set_aspect("equal")
  axs[0, 1].contour(t) 
  axs[0, 1].set_title(f"ylim={axs[0, 1].get_ylim()}")
  
  axs[1, 0].set_ylabel('origin="lower"')
  axs[1, 0].imshow(t, origin="lower")
  axs[1, 0].set_title(f"ylim={axs[1, 0].get_ylim()}")
  axs[1, 1].set_aspect("equal")
  axs[1, 1].contour(t, origin="lower")
  axs[1, 1].set_title(f"ylim={axs[1, 1].get_ylim()}")
  
  axs[2, 0].set_ylabel('origin="upper"')
  axs[2, 0].imshow(t, origin="upper")
  axs[2, 0].set_title(f"ylim={axs[2, 0].get_ylim()}")
  axs[2, 1].set_aspect("equal")
  axs[2, 1].contour(t, origin="upper")
  axs[2, 1].set_title(f"ylim={axs[2, 1].get_ylim()}")
  ```
</details>

### 1. The default behavior differs:

#### image:
- `Z[0,0]` is in upper left corner
- yrange is index-range +0.5 on both sides (-> semantics: Z is interpreted as pixel, i.e. each data point covering an area)

#### contour:
- `Z[0,0]` is in lower left corner
- yrange is index-range (-> semantics: Z is interpreted as grid of points, i.e. each data point only defines the value at that position)

#### Consequence
The different yrange behavior comes from a different interpretation of the data (points vs. pixels). These different interpretations are IMHO reasonable for the respective plot type:
- image: values are positioned at integer coordinates 0..N-1 and the yrange is half a pixel larger on both sides: -0.5..N-0.5
- contour: values are positioned at integer coordinates 0..N-1 and the yrange is limited to that. An extension to -0.5..N-0.5 would not be reasonable: unlike a pixel where the +0.5 is a canonical "constant" extrapolation due to pixels defining an area, we'd need to really extrapolate the data.

Both data interpretation and position of `Z[0, 0]` are different between contour and image. Since the defaults are so fundamental, we'll not be able to change them. We can only document the differences.

### 2. origin="lower"
In both cases, `Z[0, 0]` is in the lower left corner. -> consistent

`contour()` behaves as documented ("Z[0, 0] is at X=0.5, Y=0.5 in the lower left corner."). But the yrange (0.5, 9.5) is unreasonable: The data point distance is correctly dy=1. But the offset puts the `Z[0, 0]` value at y=0.5. When interpreting data as grid as in the default case, one would still expect y=0. It seems that y=0.5 was the attempt to adapt it to pixels (put it in the center of [0,1]), but unfortunately this is not how we place our pixels in `imshow`. `Z[0, 0]` is still at y=0 here.

#### Consequence

The correct behavior would be: "Z[0, 0] is at X=0, Y=0 in the lower left corner."

### 3. origin="upper"
In both cases, `Z[0, 0]` is in the lower left corner. -> consistent

Again, `contour()` behaves almost as documented ("Z[0, 0] is at X=N+0.5, Y=0.5 in the lower left corner."). Note that the actual position X=N-0.5 (not +, but that's a documentation issue N+0.5 would be even more weird. Nevertheless,

- as in the "lower" case positioning at half intergers does not make sense. The yrange should still be 0..N.
- While documented, 
- moving `Z[0, 0]` away from y=0 is surprising, if not outright wrong, even disregarding the inconsistent behavior wrt imshow.

#### Consequence

The correct behavior would be: "Z[0, 0] is at X=0, Y=0 in the upper left corner."
